### ISSUE TITLE
Allow using the ZFS backend in containers

### ISSUE THREAD

**Issue by pstch** [lxc/lxd#4184]
# Required information

 * Distribution: Debian GNU/Linux
 * Distribution version: stretch (9.0)
 * The output of "lxc info":
```
config:
  core.https_address: '[::]:8443'
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses:
  - 51.15.190.119:8443
  - '[2001:bc8:6005:121:208:a2ff:fe0c:37d0]:8443'
  - 10.207.241.1:8443
  - '[fd42:f66f:621f:6e0b::1]:8443'
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFbTCCA1WgAwIBAgIQFhXAK8xooqSqtRME3aaGnDANBgkqhkiG9w0BAQsFADA9
    MRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMR0wGwYDVQQDDBRyb290QG1h
    bG9uLmxlYmliLm9yZzAeFw0xNzEyMTQxNjE5MTNaFw0yNzEyMTIxNjE5MTNaMD0x
    HDAaBgNVBAoTE2xpbnV4Y29udGFpbmVycy5vcmcxHTAbBgNVBAMMFHJvb3RAbWFs
    b24ubGViaWIub3JnMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEApHDB
    k6EiIpWnrqTCPyB8itgkpyqiW9frmK4Cve8ugDtnF4N3bCeQq09ywkhIpBksT7EB
    kKY4H0DucgLT+2HPscHrBWKyW+prlWOqVvvVvVfttzqcdAo5ym8LkQHnijGd2Chb
    YOML2kNLUN5+11AfBc4pm8BxvmQ3Sh2aTmt8SUTg0HLQ0YICgUE64IU/inYEe5qG
    hmR2cKTqE78tEEekMiZVte86DVEu2/9q+OiVfEo3dty8m+PLv6fcwTH3qgJiM27s
    YwNsUm9/Cvh3s3NkvrFyl6TOj1cYhvXBI4B7rmeDx5b4s+yYccfBmAYX5R3YF0bP
    amJLhNTlg1bkmArAzK5nlxK4M3mHHuVfJkbx/PRliQVRJ83iCthyhRax0jDjx4Zc
    fnO571kt3JJY8aLT+rcagPpWlefAM0AqdWJxbXQBqkfR3xxdX1LzclEEUevyeAti
    alJVlP70f0ln+NhZxprjj/ZQuUdB9h/NTD7RtbcDS7V3J5VJ/Nmr+nHGvC/xZvbs
    HTSTVvN2kkwQu1gp5I+Qnxr6V7MdW7tipUj3hrXA476s2uk+Cg9NbhrhKsrhKJX0
    TaO0SpAE2A++ftXCffkymOiEEdP2bYgz32KO6mKan/6qDP2DIv5EM4cOR8Js1ty5
    qqtIMbpsE46TSIn00jwtfPdFQuN8c+ZePN0AUU0CAwEAAaNpMGcwDgYDVR0PAQH/
    BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwMgYDVR0R
    BCswKYIPbWFsb24ubGViaWIub3JnhwQzD753hxAgAQvIYAUBIQIIov/+DDfQMA0G
    CSqGSIb3DQEBCwUAA4ICAQApdor9j+xiu1BzAFCI8Enm1J7bIaY11MctycgEtYAk
    G6WYkuItqLsuCVACaow1dtwfN1uU2I6Z07aGIxwNo75ghrY4OesXsM8ymmBSG8Z+
    vxZ3+ixvxMbTdzL0pHY7K6aHQMHAnPnb6hMgl/SbXj8tOaah3sOam8fvqoChLv20
    koCro8fNmjYF6zpoRL/a0MwJuOu/iBWLxK1cq4JmSGE9tn5WjVGXq2+OUHA5KAF7
    EZHRMAN9jqRIWdFlphmkhtRO2/SHplV2xqXE7lU/dhBFT0+p+ndsZfK4Y9a9M6ed
    or0lHIuUsRRHfY7w2P/jWWyT8zWzIi9YnV/cV9+LzSwQiK5X5eQacPj7sFTATAsN
    LPu7kFfX5b4wtG5rgZxR8g9iWsgRhUXH0pgTpqSsvwuvePMOIxuB2TTxwnGUH3im
    ReKnWhMMGkXSo4mqM7n4LgoFUowau3QIqIApL9FlCsLYTt4iV7IrvgRmY6ZDuosU
    Tosdv3a5V/yWgxJvuoeAVQMGTo7QqxhdAT/MvQpqXuZv433rl/TCC0hXmsELl2o9
    9TsJD6OwZ9G0t/N+fvGWGY2O1y/6bkaaMGQ1TQfIYFNi3tC3ZYBiQRBk01aVcxrt
    7+pIvtOUs99IhE22tGsrgm2NJ86UJ+NemEvQdUQGR8UfO/XVCKR45ySR3yTE9pjN
    lg==
    -----END CERTIFICATE-----
  certificate_fingerprint: 7dbf89a574102eecc363c8b78d9ea0862f6a0779d92fb94a78fd4b35a8f2b42e
  driver: lxc
  driver_version: 2.0.7
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.9.0-4-amd64
  server: lxd
  server_pid: 27993
  server_version: "2.18"
  storage: zfs
  storage_version: 0.8.0-test
```
# Issue description

The ZFS backend can only be used in containers if the ZFS module is installed in the container (having the ZFS utilities installed is not sufficient). It should be possible to use the ZFS backend from a container that doesn't have the module installed, as long as the module is properly loaded from the host.

# Steps to reproduce

 1. Install ZFS and LXD and create an unprivileged container with access to `/dev/zfs`
 2. Install the ZFS utilities (but not the modules) in the container 
 3. Attempt to run `lxd init` in the container
 4. Observe that it doesn't allow ZFS as a backend
 5. Install `zfs-dkms` and `spl-dkms` in the container
 6. Observe that ZFS is now an allowed backend

# Additional information

I think that this is caused by the ZFS backend attempting to load the `zfs` module in its initialization code, which fails from a container that doesn't have the modules installed. Having to compile the modules in each container is a bit cumbersome, so maybe this module load operation could be made optional ?

**Comment by stgraber** [comment]
This is not possible, ZFS isn't aware of namespaces at all.

Allowing access to /dev/zfs in your container will result in the various zfs tools showing you your zpools and zfs datasets of the host, leading to obvious dataloss and breakages.

That is why ZFS is clearly listed as not supporting nesting in our storage documentation.

We'd obviously love for ZFS to actually gain full container support, making it safe to use in such an environment, until then, we'll stick to not allowing its use and would strongly recommend users not try to workaround it :)

**Comment by pstch** [comment]
Upcoming work in ZFS should make it aware of user namespaces (see [this PR](https://github.com/zfsonlinux/zfs/pull/6865)), and I'm currently testing this experimental setup (using `zfs allow/zfs unallow` to restrict access to containers) with relative success.

There are some other limitations and problems (prohibiting the host from mounting the container's datasets, for example, and the ability to allow UID ranges rather than individual UIDs), but they can be mostly worked around for now.

I understand recommending users not try to workaround it. I can stick to building the unnecessary modules in the containers until ZFS has more proper user namespaces support.

**Comment by stgraber** [comment]
Yeah, once zfs upstream has this merged and it's functional for LXD's use case, we'll tweak the detection code to allow it if all the right bits are at the right version, until then, you'll have to workaround our checks I think :)

**Comment by stgraber** [comment]
I'm subscribed to the ZFS PR and we have @blub on our IRC channel already so I'm sure we'll be kept up to date on this :)

**Comment by pstch** [comment]
OK, thanks. Sorry for the noise.

**Comment by samy-mahmoudi** [comment]
New relevant PR at OpenZFS:

- https://github.com/openzfs/zfs/pull/12263

**Comment by M1cha** [comment]
I got merged in https://github.com/openzfs/zfs/commit/4ed5e25074ffec266df38556d9b3a928c5e0dee9  :tada: 
So what to we have to do to use this (safely) with LXD? Can we simply give the container access to `/dev/lxd` and update the documentation or do we need any changes to the code?

**Comment by stgraber** [comment]
You can try giving /dev/zfs access through a unix-char device. For us to do it by default, we're going to want to see a stable release with it and possibly wait for a point release on it so any crazy security issue should have been found by then.

Passing /dev/zfs on a system which doesn't support usage inside containers is a very very bad idea, so please make sure you have the correct ZFS installed before you add such a device to a container.

**Comment by M1cha** [comment]
If I understand the sample from the commit message correctly we do need explicit support though:
```bash
zfs zone /proc/1234/ns/user containers/unpriv
```
This sounds like we have to explicitly attach a volume to a namespace. Unlike you wanna do that every time you restart the container, LXD has to do that.
But that's also a good thing because LXD probably knows best which volumes should be accessible inside the container.

**Comment by michacassola** [comment]
Uhhhh, this will be great!

**Comment by stgraber** [comment]
yep!

**Comment by mklemm2** [comment]
Any news? Is anyone working on this?

**Comment by tomponline** [comment]
Its not in a zfs release afaik

**Comment by m-ueberall** [comment]
> Any news? Is anyone working on this?

See [this comment upstream](https://github.com/openzfs/zfs/commit/4ed5e25074ffec266df38556d9b3a928c5e0dee9#commitcomment-80758905) regarding the required prerequisite: “_[T]his feature isn't planned for the 2.1.x point releases, but is planned for the next major release._”

**Comment by dalbani** [comment]
I know that OpenZFS 2.2 hasn't been released yet — any guess when this will happen? — but is there a way to try this new `zone` feature in combination with LXD?
After having compiled and installed `openzfs-zfsutils` from Git on both the host and inside an LXC container, I've tried to play with commands like mentioned above:
```
host# zfs set zoned=on data/lxd/containers/container-1
host# zfs zfs zone /proc/<container-PID>/ns/user data/lxd/containers/container-1
```
But trying to use `zfs list` within the container fails:
```
container-1# strace zfs list
... 
access("/dev/zfs", F_OK)                = 0
openat(AT_FDCWD, "/dev/zfs", O_RDWR|O_EXCL|O_CLOEXEC) = -1 EPERM (Operation not permitted)
```

My goal is have MicroK8s run in an LXC container in combination with OpenEBS ZFS LocalPV. Thus the need to be able to manage datasets under `data/lxd/containers/container-1` in this example.

Any help would be appreciated!

**Comment by dalbani** [comment]
Ping!
Can anyone help me (investigate how to) use ZFS zone functionality in an LXC context? Be it with a non released OpenZFS version.
Thanks!

**Comment by stgraber** [comment]
I'd strongly recommend testing with a clean container, the microk8s container is using both nesting and privileged flags which I can't think would be a good combination for this.

**Comment by dalbani** [comment]
Use a generic container made indeed all the difference, thanks for the tip!
For the record, here's what I see on the host:
```
host $ zfs list
NAME                                                                                       USED  AVAIL  REFER  MOUNTPOINT
data                                                                                      62.3G  1.89T   185K  /data
data/lxd                                                                                  14.3G  1.89T   170K  legacy
data/lxd/buckets                                                                           170K  1.89T   170K  legacy
data/lxd/containers                                                                       12.5G  1.89T   170K  legacy
data/lxd/containers/microk8s-1                                                            11.7G  1.89T  12.5G  legacy
data/lxd/containers/zfs-1                                                                  840M  1.89T  1.68G  legacy
...
```

On the other hand, here's what it looks like from within a container:
```
host $ lxc exec zfs-1 -- bash
root@zfs-1:~# zfs list
NAME                        USED  AVAIL  REFER  MOUNTPOINT
data                       62.3G  1.89T   185K  /data
data/lxd                   14.3G  1.89T   170K  legacy
data/lxd/containers        12.5G  1.89T   170K  legacy
data/lxd/containers/zfs-1   840M  1.89T  1.68G  legacy
root@zfs-1:~# zfs create data/lxd/containers/zfs-1/test
root@zfs-1:~# zfs list
NAME                             USED  AVAIL  REFER  MOUNTPOINT
data                            62.3G  1.89T   185K  /data
data/lxd                        14.3G  1.89T   170K  legacy
data/lxd/containers             12.5G  1.89T   170K  legacy
data/lxd/containers/zfs-1        840M  1.89T  1.68G  legacy
data/lxd/containers/zfs-1/test   170K  1.89T   170K  legacy
```

And what I did beforehand was to set:
```
host # zfs set zoned=on data/lxd/containers/zfs-1
host # zfs zone /proc/3396158/ns/user data/lxd/containers/zfs-1
```
Where `3396158` was taken from `lxc info zfs-1`.

So my question is: how to go further now?
I suppose I need to dig out to see why the `microk8s-zfs` profile prevents this from working.
But I have little experience with LXC/LXD, so maybe someone more knowledgeable on this topic may somehow already know the answer?

**Comment by stgraber** [comment]
Because the zfs zone stuff is based on the user namespace, I don't expect it to ever work with security.privileged=true as that turns off the user namespace.

**Comment by stgraber** [comment]
You'd need to get microk8s working properly without security.privileged=true for any of this to work there.

**Comment by dalbani** [comment]
For the record, while I managed to build a Snap of LXD by applying some changes in https://github.com/lxc/lxd-pkg-snap for ZFS 2.1.99 as I mentioned in my previous message, installing this snap somehow failed (`lxc: 6: exec: aa-exec: Permission denied`).
Maybe it had to do with the `--devmode` flag that I had to use due to the unsigned snap?

Anyway, I switched gear and went to Debian 12 (to be very soon released as `stable`) with LXD 5.0, where I didn't have to recompile anything.
I just configured my container to run a script for the `start-host` and `post-stop` hooks, like:
```
#!/bin/sh -eu

# FIXME find a way to programatically find the parent dataset name — any suggestion?
ZFS_DATASET=default/containers/${LXC_NAME}

if [ ${LXC_HOOK_TYPE} = start-host ]; then
  zfs set zoned=on ${ZFS_DATASET}
  zfs zone /proc/${LXC_PID}/ns/user ${ZFS_DATASET}
elif [ ${LXC_HOOK_TYPE} = post-stop ]; then
  zfs set zoned=off ${ZFS_DATASET}
fi
```

And, voilà, ZFS delegation in the LXC container works just fine!
Well, I already saw a (small) issue where editing the configuration of a container while it's running (and thus its ZFS dataset has been zoned) is not possible.
Anyway, no big deal for me, I just stop the container first and then edit the configuration if I need to.

**Comment by mdsitton** [comment]
We now have https://github.com/openzfs/zfs/releases/tag/zfs-2.2.0-rc1
